ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -122,119 +123,223 @@ hexDump :: BL.ByteString -> String -> IO ()
 hexDump buf out | BL.empty == buf = say out
 hexDump buf out = hexDump (BL.tail buf) (out ++ printf "0x%02x " (BL.head buf))
 
+
 -- |
--- Run a node using @'NetworkInterface'@ using a socket.  It will start to
--- listen on incomming connections on the supplied @'nodeAddress'@, and returns
--- @'NetworkNode'@ which let one connect to other peers (by opening a new
--- TCP connection) or shut down the node.
---
--- When connecting to a remote node using @connectTo@ th
-withNetworkNode
-  :: forall ptcl t r.
+-- Connect to a remote node.  It is using bracket to enclose the underlying
+-- socket acquisition.  This implies that when the continuation exits the
+-- underlying bearer will get closed.
+connectTo
+  :: forall ptcl.
      ( Mx.ProtocolEnum ptcl
      , Ord ptcl
      , Enum ptcl
      , Bounded ptcl
      , Show ptcl
      , Mx.MiniProtocolLimits ptcl
      )
-  => NetworkInterface ptcl Socket.AddrInfo IO r
-  -> (NetworkNode Socket.AddrInfo IO r -> IO t)
-  -> IO t
-withNetworkNode NetworkInterface {nodeAddress, protocols} k =
-    bracket mkSocket Socket.close $ \sd -> do
+  => MuxApplication InitiatorApp ptcl IO
+  -- ^ application to run over the connection
+  -> Socket.AddrInfo
+  -- ^ local address; the created socket will bind to it
+  -> Socket.AddrInfo
+  -- ^ remote address
+  -> IO ()
+connectTo app localAddr remoteAddr =
```

Comment:
If I wish to `connectTo` a peer, I must give a `MuxApplication InitiatorApp ptcl IO`. I have to pick a `ptcl`. But if I'm not mistaken (very well could be) the `ptcl` is peculiar to a _version_, which is supposed to be determined by the handshake protocol. How can I use `connectTo` _and_ the handshake protocol?

Reply:
`ptcl` is outside of version negotiation; there will be a single `ptcl` on a given port.  `ptcl` indicates the overall protocol version, handshake is only for negotiation it's parameters.  For using `connectTo` with handshake protocol see the [coot/handshake-negotiation-2](https://github.com/input-output-hk/ouroboros-network/blob/coot/handshake-integration-2/ouroboros-network/src/Ouroboros/Network/Socket.hs#L181) branch.